### ISSUE TITLE
Startup fails with --storage.tsdb.retention < 20h: Opening storage failed: at least one block-range must exist

### ISSUE THREAD

**Issue by chlunde** [prometheus/tsdb#122]
version=2.0.0-beta.1

```
/opt/prometheus-2.0.0-beta.1.linux-amd64/prometheus --storage.tsdb.retention=19h
INFO[0000] Starting prometheus (version=2.0.0-beta.1, branch=HEAD, revision=4dcb465029f39a2a3030333a81753f8a40ec90ce)  source="main.go:204"
INFO[0000] Build context (go=go1.8.3, user=root@d3f9974fac5a, date=20170811-12:12:36)  source="main.go:205"
INFO[0000] Host details (Linux 3.10.0-693.el7.x86_64 #1 SMP Thu Jul 6 19:56:57 EDT 2017 x86_64 coffeehouse.test.sparebank1.no (none))  source="main.go:206"
INFO[0000] Starting tsdb                                 source="main.go:218"
ERRO[0000] Opening storage failed: at least one block-range must exist  source="main.go:221"
```

It works with 20h and higher.  Looks like it is returned from https://github.com/prometheus/tsdb/blob/c4ca881685ae1266a75caf57da46d8b6934213c0/compact.go#L33-L43

Should this "just work", or require tuning from the user?

**Comment by fabxc** [comment]
Thanks for reporting. Definitely a case we should be handling. The function here should return the right thing. Just that we are truncating the result later on to apply a maximum block size w.r.t to retention time:

https://github.com/prometheus/prometheus/blob/dev-2.0/storage/tsdb/tsdb.go#L61-L68

I was confused why it happens at 19h already. Turns out the `10` and `3` arguments in the call to `ExponentialBlockRanges` are flipped – that's actually relatively bad.

@Gouthamve

**Comment by fabxc** [comment]
Closing here since it's an integration issue in Prometheus.

Fix in prometheus/prometheus#3089

**Comment by cboggs** [comment]
Sorry to comment on a closed issue, but there's a workaround here if you'd like (until the fix is merged into P8s). This set of args worked for me when I wanted to run a tiny short-retention instance:
```
--storage.tsdb.min-block-duration=15m
--storage.tsdb.max-block-duration=30m
--storage.tsdb.retention=1h
```

This way the retention period encompasses at least one block range, and thus is stops throwing the error and crashing. :-)
